polar-core/src/sugar.rs

```diff
@@ -0,0 +1,850 @@
+    relations: Option<Term>,
+) -> HashMap<Term, Declaration> {
+    // Fold List<role> => HashMap<role, Declaration>
+    let declarations = roles
+        .into_iter()
+        .flat_map(|inner| inner.value().as_list().unwrap().clone())
+        .fold(HashMap::new(), |mut acc, role| {
+            acc.insert(role, Declaration::Role);
+            acc
+        });
+
+    // Fold List<permission> => HashMap<permission_or_role, Declaration>
+    let declarations = permissions
+        .into_iter()
+        .flat_map(|inner| inner.value().as_list().unwrap().clone())
+        .fold(declarations, |mut acc, permission| {
+            acc.insert(permission, Declaration::Permission);
+            acc
+        });
+
+    // Fold Dict<relation, resource> => HashMap<permission_or_role_or_relation, Declaration>
+    relations
+        .into_iter()
+        .flat_map(|inner| inner.value().as_dict().unwrap().fields.clone())
+        .fold(declarations, |mut acc, (relation, resource)| {
+            acc.insert(
+                resource.clone_with_value(value!(relation.0.as_str())),
+                Declaration::Relation(resource),
+            );
+            acc
+        })
+}
+
+fn resource_as_var(resource: &Term) -> PolarResult<Value> {
+    let name = &resource.value().as_symbol()?.0;
+    let mut lowercased = name.to_lowercase();
+
+    // If the resource's name is already lowercase, append "_instance" to distinguish the variable
+    // name from the resource's name.
+    if &lowercased == name {
+        lowercased += "_instance";
+    }
+
+    Ok(value!(sym!(lowercased)))
+}
+
+fn rewrite_implier_as_rule_body(
+    implier: &Term,
+    relation: &Option<Term>,
+    namespaces: &Namespaces,
+    resource: &Term,
+) -> PolarResult<Term> {
+    let resource_var = implier.clone_with_value(resource_as_var(resource)?);
+    let actor_var = implier.clone_with_value(value!(sym!("actor")));
+    if let Some(relation) = relation {
+        // TODO(gj): what if the relation is with the same type? E.g.,
+        // `Dir { relations = { parent: Dir }; }`. This might cause Polar to loop.
+        let related_type = namespaces.get_related_type(resource, relation)?;
+        let related_resource_var = relation.clone_with_value(resource_as_var(related_type)?);
+
+        let relation_call = relation.clone_with_value(value!(Call {
+            name: sym!("relation"),
+            // For example: vec![org, "parent", repo]
+            args: vec![related_resource_var.clone(), relation.clone(), resource_var],
+            kwargs: None
+        }));
+
+        let implier_call = implier.clone_with_value(value!(Call {
+            name: namespaces.cross_resource_predicate_name(resource, relation, implier)?,
+            // For example: vec![actor, "owner", org]
+            args: vec![actor_var, implier.clone(), related_resource_var],
+            kwargs: None
+        }));
+        Ok(implier.clone_with_value(value!(op!(And, relation_call, implier_call))))
```

Comment:
Users won't ever see the expanded version (even when debugging), so the order probably doesn't matter much. The debugger would visit the expanded `relation` call first and then the expanded `implier` call, but since we store the source info from the original shorthand implication they won't actually see the expanded forms.

Reply:
Made a note to follow up on this.